reference/solver/batch_cg_kernels.hpp

```diff
@@ -96,7 +96,7 @@ inline void update_x_and_r(
     batch_single_kernels::compute_conj_dot_product_kernel<ValueType>(
         p_entry, Ap_entry, alpha_entry);
 
-    const ValueType temp = rho_old_entry.values[0] / alpha_entry.values[0];
+    const ValueType temp = rho_new_entry.values[0] / alpha_entry.values[0];
```

Comment:
I dont think this is accurate. In this case here, `rho_old` is $<r_{j}, z_{j}>$, while `rho_new` is $<r_{j+1}, z_{j+1}>$. And `temp` here is equivalent to $\alpha_j$ in Algorithm 9.1 in Saad's book for example. So I think the current state in `develop` should be correct.

Reply:
But we are passing `rho_new_entry` to this function. Maybe `rho_new` means $<r_j, z_j>$ and `rho_old` means $<r_{j-1}, z_j>$?

Reply:
No, I think the order of the function calls needs to be re-arranged. From looking at Saad's book, I think here is the diff that should be correct. The CUDA/HIP/DPCPP kernels should be correct. 

```diff
diff --git a/reference/solver/batch_cg_kernels.hpp b/reference/solver/batch_cg_kernels.hpp
index d4a35e3d0..2c0309478 100644
--- a/reference/solver/batch_cg_kernels.hpp
+++ b/reference/solver/batch_cg_kernels.hpp
@@ -199,6 +199,10 @@ inline void batch_entry_cg_impl(
                  gko::batch::to_const(rho_old_entry),
                  gko::batch::to_const(z_entry), p_entry);
 
+        // rho_old = rho_new
+        batch_single_kernels::copy_kernel(gko::batch::to_const(rho_new_entry),
+                                          rho_old_entry);
+
         // Ap = A * p
         batch_single_kernels::simple_apply(
             A_entry, gko::batch::to_const(p_entry), Ap_entry);
@@ -207,12 +211,8 @@ inline void batch_entry_cg_impl(
         // x = x + temp * p
         // r = r - temp * Ap
         update_x_and_r(
-            gko::batch::to_const(rho_new_entry), gko::batch::to_const(p_entry),
+            gko::batch::to_const(rho_old_entry), gko::batch::to_const(p_entry),
             gko::batch::to_const(Ap_entry), alpha_entry, x_entry, r_entry);
-
-        // rho_old = rho_new
-        batch_single_kernels::copy_kernel(gko::batch::to_const(rho_new_entry),
-                                          rho_old_entry);
     }
 
     logger.log_iteration(batch_item_id, iter, res_norms_entry.values[0]);
```

Reply:
No, I think the original is correct.
the implementation is rotated a bit because we put the criterion right after updating r.
we move the rest after updating r in the begining of the loop by different initial value of p.

Use preconditioned algorithm in https://en.wikipedia.org/wiki/Conjugate_gradient_method 
![image](https://github.com/user-attachments/assets/98c6e805-9375-4150-ad55-113c6e66b6e9)
we move the code after `if` to the beginning, so we do not need preconditioner and p = z out of loop in the picture by setting the p = 0 -> we get p = z in the first iteration.

Reply:
This function is called with `rho_new_entry`. This is the same as for the non-batched CG, and it matches @yhmtsai explanation. The cuda/hip version is differently ordered, since the initialization is different (`rho_old` is set to $<r_0, z_0>$ instead of `1`).
So I think calling the function with `rho_new_entry` is correct, and the parameter name should be changed here.

Reply:
Yes, you are right. Disregard my comments.